Set aspect ratio preserving CSS on embedded content with fixed size iframes

notnownikki opened the pull request:
## Description

To support the wide styles, we need to preserve the aspect ratio of fixed size embeds when they are resized to full width.

This applies a css class to embedded content that uses iframes (e.g. YouTube, Vimeo) to preserve aspect ratio.

## How has this been tested?

Embed the following videos:

https://www.youtube.com/watch?v=XT13ijUfSts

https://www.youtube.com/watch?v=uAE6Il6OTcs

The first should have `wp-embed-aspect-16-9` applied, and the second should have `wp-embed-aspect-4-3`.

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

notnownikki committed:
Set aspect ratio preserving CSS on previews with fixed size iframes

Commit:
Add aspect ratio CSS.

notnownikki commented:
> I also put this in a themes.scss file, as an "opinionated style". That means a theme has to opt into this. Your thoughts here are welcome, I'd be fine with moving it to style.scss as it's a nice feature to have for the stock offering.

I think this should be in `style.scss` so that no matter what the theme has as its content size, the embeds will always fit as they should.

> Can you make it so that same wrapper that exists in the editor, exists on the frontend? Doesn't have to be for all embeds (but can be), just ones that need to be responsive (i.e. video).

I'll look at it now :)

>Eventually if we can get this working, should we consider removing the onResize responsiveness that exists for embeds right now? The idea being that a CSS only solution might be more performant?

This is exactly what I want to do!

notnownikki committed:
Add embed wrapper div in saved content for responsive classes

notnownikki commented:
@jasmussen I added the wrapper div and added a block migration so that existing embeds will update when the posts are edited. There seem to be some lint issues with the new css though,

```
0] [2] packages/block-library/src/embed/theme.scss
[0] [2]  12:4   âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  29:24  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  33:24  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  37:24  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  41:23  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  45:23  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  49:23  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
[0] [2]  53:23  âœ–  Expected double colon pseudo-element notation   selector-pseudo-element-colon-notation
```

Would you be able to fix those if you move the css over to `style.scss`?

jasmussen committed:
Fix things up.

jasmussen commented:
Behold, our magic at work:

![behold](https://user-images.githubusercontent.com/1204802/45030677-f0dfc500-b04c-11e8-9058-6918754d25d7.gif)

Isn't that amazing? I think it's amazing. That's Twenty Seventeen, which has no concept of blocks.

🤘

> There seem to be some lint issues with the new css though,

I could swear I fixed those. They blocked my commit and I had to fix them before I could push. So I'm confused why they regressed. But I've fixed them again in https://github.com/WordPress/gutenberg/pull/9500/commits/372078486c596e06eedd947945b2c9e634d661fb

notnownikki committed:
Remove the aspect ratio preservation from sandbox

notnownikki commented:
@jasmussen  Last commit removes the aspect ratio preservation javascript from the sandbox. It's working brilliantly on the front end, but in the editor there's something throwing it off and I'm not sure what. If you embed a 16:9 video you'll see it's not as high as it needs to be in the editor. Perhaps the caption editable is overlapping somehow?

notnownikki committed:
Fix embed tests

jasmussen committed:
Fix for responsiveness in the editor also.

Note that this changes the sandbox component a little bit. We want to make sure that the changes we made here don't negatively affect other things that use the same component.

jasmussen commented:
Okay I pushed a fix:

![behold again](https://user-images.githubusercontent.com/1204802/45035026-47530080-b059-11e8-8b36-ec9912b902ff.gif)

However as noted in the commit note, I had to make a change to the sandbox component in order to make this work. But I'm not sure where else that component is being made — for example if it's being used for embeds that are not supposed to be responsive, the height property I added is undesirable. Do we have a way to scope this a bit so only videos receive the height stuff?

jasmussen commented on packages/components/src/sandbox/index.js:
```diff
@@ -137,8 +137,12 @@ class Sandbox extends Component {
+			body > div,
 			body > div > iframe {
 				width: 100%;
+				height: 100%;
```
Note that this height property is necessary in order for the responsiveness to work, basically because every element that contains the embed needs an explicit height in order for a single percentage height to work properly. 

But if this component is used for things that aren't aspect ratio responsive, we need to scope it. Something like:

```
html.is-video,
html.is-video body,
html.is-video body > div,
html.is-video body > div > iframe {
height: 100%;
}
```

notnownikki replied:
Ah, I see :) Ok, I'll see how it can be changed around to only get applied to things we're doing aspect ratio preservation on.

jasmussen replied:
Feel free to send this PR right back at me if we can have a CSS class or separate style declaration for aspect ratio responsive stuff only.

chrisvanpatten replied:
It would be cool to have an `.is-responsive` or `.has-aspect-ratio` class to target here.

notnownikki replied:
Agreed. I was thinking `.wp-embed-has-aspect-ratio` and then the editor should be able to use that to apply the height to the sandbox iframe.

notnownikki committed:
Fix height for video sandboxing

notnownikki committed:
Make the aspect ratio generic again

jasmussen commented on packages/components/src/sandbox/index.js:
```diff
@@ -142,6 +142,11 @@ class Sandbox extends Component {
 			body > div,
 			body > div > iframe {
 				width: 100%;
+			}
```
Gloriously done. I haven't tested, but based on looking at the code this should work. 👍 👍

jasmussen commented:
If this is ready for review, would you mind extending the review range to for example gutenberg-core, and perhaps assigning a 3.9 milestone?

talldan commented on packages/block-library/src/embed/index.js:
```diff
@@ -156,6 +178,47 @@ export function getEmbedEdit( title, icon ) {
+			// calculate the aspect ratio and set an extra css class so that the rendered
+			// content keeps the correct height no matter how wide the block is set to be.
+			const { height, width } = this.getiFrameHeightWidth( html );
+			if ( undefined !== height && undefined !== width ) {
```
I wonder if most of this logic could be moved into `getIframeHeightWidth` and that could become `getAspectRatioClassName` so that `setAttributesFromPreview` isn't doing so much. Or they could be separate functions.

talldan commented on packages/block-library/src/embed/index.js:
```diff
@@ -156,6 +178,47 @@ export function getEmbedEdit( title, icon ) {
+
+				if ( aspectRatioClassName ) {
+					aspectRatioClassName += ' wp-has-aspect-ratio';
+					const className = classnames( this.props.attributes.className, aspectRatioClassName );
```
Could avoid the concatenation here by making classnames do the work:
```js
const className = classnames( this.props.attributes.className, 'wp-has-aspect-ratio', aspectRatioClassName );
```

talldan commented on packages/block-library/src/embed/index.js:
```diff
@@ -218,6 +281,7 @@ export function getEmbedEdit( title, icon ) {
 			const cannotPreview = includes( HOSTS_NO_PREVIEWS, parsedUrl.host.replace( /^www\./, '' ) );
 			// translators: %s: host providing embed content e.g: www.youtube.com
 			const iframeTitle = sprintf( __( 'Embedded content from %s' ), parsedUrl.host );
+			const sandboxClassnames = className ? type + ' ' + className : type;
```
Could use classnames again here, as it'll ignore className when its undefined:
```
const sandboxClassnames = classnames( type, className );
```

talldan reviewed (changes requested):
This works really nicely, great work 😁

I've added a few minor comments.

notnownikki commented:
@talldan thank you! I'll make all those changes. `querySelector` has slipped my mind and it would do exactly what we need.

notnownikki committed:
Fix review points

notnownikki commented:
Ok, think I fixed all those points. I moved all of the aspect ratio CSS code into `maybeSetAspectRatioClassName`.

talldan commented:
Looks good to me. Thanks for making those changes. Looks like codecov has failed, not sure what can be done about that.

aduth commented on packages/block-library/src/embed/index.js:
```diff
@@ -136,6 +137,58 @@ export function getEmbedEdit( title, icon ) {
+		 * @param {string} html The preview HTML that possibly contains an iframe with width and height set.
+		 */
+		maybeSetAspectRatioClassName( html ) {
+			const previewDom = document.createElement( 'div' );
```
Minor: DOM is an acronym and as such should be capitalized as `previewDOM` ([reference](https://github.com/WordPress/gutenberg/blob/master/docs/reference/coding-guidelines.md#variable-naming))

aduth commented on packages/components/src/sandbox/index.js:
```diff
@@ -176,7 +158,7 @@ class Sandbox extends Component {
 		// we can use this in the future to inject custom styles or scripts
 		const htmlDoc = (
-			<html lang={ document.documentElement.lang }>
+			<html lang={ document.documentElement.lang } className={ this.props.type }>
```
What is the `type` prop meant to represent?

aduth commented on packages/block-library/src/embed/index.js:
```diff
@@ -136,6 +137,58 @@ export function getEmbedEdit( title, icon ) {
+		 */
+		maybeSetAspectRatioClassName( html ) {
+			const previewDom = document.createElement( 'div' );
+			previewDom.innerHTML = html;
```
I am very wary of the security implications of this line. This will allow JavaScript in the preview markup to be evaluated, even though we're never including it in the DOM. As a demonstration, paste the following into your console while viewing the editor:

```js
document.createElement( 'div' ).innerHTML = '<img src=/ onerror=\'alert("haxd")\'>'
```

Now it's a question as to whether we consider embed preview markup to be "safe". Given that the Sandbox component exists, I'm operating on the assumption that the answer is **no**. Therefore, this is a security vulnerability if it comes to be released.

notnownikki replied:
Changed this to use a regex in https://github.com/WordPress/gutenberg/pull/9770

aduth replied:
Another simpler option might be `document.implementation.createHTMLDocument` as a "sandbox" of sorts.

Example: https://github.com/aduth/hpq/blob/f17b3fdc9c5692e9f676f94c33a003d191c81bd6/src/index.js#L21-L23

aduth commented:
The demo content should have been updated here. While the deprecation migration saves it from being flagged as invalid, it does cause the demo post to trigger an autosave after delay.

```
validation.js:150 Block validation: Block validation failed for `core-embed/vimeo` ({name: "core-embed/vimeo", title: "Vimeo", description: "Add a block that displays content pulled from other sites, like Twitter, Instagram or YouTube.", icon: {…}, category: "embed", …}).

Expected:

<figure class="wp-block-embed-vimeo alignwide wp-block-embed is-type-video is-provider-vimeo"><div class="wp-block-embed__wrapper">
https://vimeo.com/22439234
</div></figure>

Actual:

<figure class="wp-block-embed-vimeo wp-block-embed alignwide is-type-video is-provider-vimeo">https://vimeo.com/22439234</figure>
```

chrisvanpatten commented:
There's a report in #10109 about this CSS causing a conflict with a theme that's doing its own responsive video handling.

It's only in one theme so far (that we know of) so maybe just something the theme dev needs to patch, but if there's a more widespread practice of themes including some kind of responsive video functionality, it may be worth revisiting a way to make this more robust or opt-in.
